### PR TITLE
Fix memory leak

### LINKED ISSUE #102: Memory leak when required repeatedly
(This has been my first effort at debugging a memory leak, so if you have any advice on how I can do this better or be more helpful, please let me know!)

I think I've run into a memory leak in this module. Here's the short backstory: I work on a project, [stylelint](https://github.com/stylelint/stylelint), that has *a lot* of tests. We end up running >20K assertions across hundreds of test files. We've been experimenting with the test runner Jest, but it always crashes on Node 4. I logged heap statistics and with each test file saw the heap growing, with inadequate cleanup, until the crash. Through some trial-and-error experiments, I think I tracked this behavior down to a dependency, [cosmiconfig](https://github.com/davidtheclark/cosmiconfig), that uses *this* module. When I change cosmiconfig's usage of `graceful-fs` to Node's `fs`, stylelint's memory problem goes away: the garbage collection adequately reduces memory usage when the heap has grown too much.

I'm not sure what the next steps would be for diagnosing this, but I can try to help if anybody has any suggestions.

### PR DESCRIPTION
**Pull request by dylang**
`graceful-fs` is relied on by many major projects. Test frameworks, such as `jest`, both reset the module cache between tests to facilitate mocking, and rely on `graceful-fs`. 

Because of this, `graceful-fs` is re-wrapping `fs` multiple times, causing ever-growing memory leaks. In a project with a a few thousand tests, this balloons to GB.

Three memory leaks were fixed in this PRz
* `fs.close` was getting re-wrapped every time `graceful-fs` was loaded.
* `fs.closeSync` was getting re-wrapped every time `graceful-fs` was loaded.
* The same clone of `fs` generated by `./fs.js` was getting re-wrapped every time `graceful-fs` was loaded.

These memory leaks show up when:
* More than one version of `graceful-fs` is required.
* The module cache is cleared, which happens in Jest and other test frameworks.
* The impact of this leak is multiplied as [thousands of modules](https://www.npmjs.com/package/graceful-fs) rely on this package.

Note:
`--enable-gc` was added to the test runner so that we can monitor memory usage in one of the tests.

Fixes #102 

Some related tickets:
* https://github.com/facebook/jest/issues/7274
* https://github.com/facebook/jest/issues/6738
* https://github.com/facebook/jest/issues/6814
* https://github.com/facebook/jest/issues/6399
* https://github.com/mapbox/underreact/issues/15
* https://github.com/stylelint/stylelint/issues/2168
* https://github.com/niftylettuce/email-templates/issues/318

### CODE REVIEW

**Comment by tmax** [comment]
this is cool, but i have a noobish question.  how do you test or "measure" the memory-leak-fix? 

do you have a script to simulate typical usage and compare the expected memory usage vs actual?  
any tools for doing that?  interested to know how you'd solve this type of problem in general (tooling for fixing memory leaks).

**Comment by dylang** [comment]
Hi @tmax - great question!

If you look in `test/avoid-memory-leak.js`, you'll see we're using [importFresh](https://github.com/sindresorhus/import-fresh) from the always amazing @sindresorhus to load the module. 

Normally, node caches modules once they are `required`, but this library deletes the cache before loading so that we can reload it. This simulates what happens in frameworks like `Jest` that clear the module cache between tests. 

First, we look at the memory usage with `process.memoryUsage().heapUsed / Math.F(1024, 2);` The math is to convert it to Mb for easy reading.

Then we require "fresh" the module 4000 times, simulating our 4800 unit tests.

We then call `global.gc();` - this is an internal node command garbage collect. We normally don't need this, node will garbage collect as it has time, but in this 
case we want to measure the memory impact right away.

Before this fix, garbage collecting had no effect because every copy of the module was still referenced by `fs`.

Finally, we look at the memory usage with `process.memoryUsage().heapUsed / Math.pow(1024, 2)` again, and compare to what we had before. A little growth is natural, so I made sure it's less than 2mb. Without the fix, the number can be in the hundreds of MB.

**Comment by isaacs** [comment]
Thanks!  Landed on graceful-fs@4.1.12.

File: graceful-fs.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
 var fs = require('fs')
 var polyfills = require('./polyfills.js')
 var legacy = require('./legacy-streams.js')
+var clone = require('./clone.js')
```
Comment (gxcsoccer) [inline graceful-fs.js]: missing clone.js

**Comment by kirillgroshkov** [comment]
great work!

**Comment by addaleax** [comment]
ping @dylang ?

**Comment by DanielGibbsNZ** [comment]
This works fine on version 4.1.15 for fixing the memory leak, but when I upgrade to version 4.2.3 the memory leak comes back. Was a regression introduced with the changes in 4.2.x?

**Comment by dylanwulf** [comment]
I'm experiencing the same as @addaleax. I have this package locked at v4.1.15 but it's still causing memory leaks in node v11, v12, v13

**Comment by DanielGibbsNZ** [comment]
Using `graceful-fs` 4.2.4 on Node 12.19.1 seems to be working fine without any noticeable memory leak.

**Review by manjaneqx — COMMENTED** [review]
🔌

**Pull request merged**